core/tests/unit/kmnkbd/test_kmx_xstring.cpp

```diff
@@ -1268,6 +1268,56 @@ test_utf32() {
   assert(s1 == std::u16string(u"🙀"));
 }
 
+void
+test_u16string_to_u32string() {
+  // normal cases
+  {
+    const std::u32string str = u16string_to_u32string(u"");
+    assert_equal(str.length(), 0);
+  }
+  {
+    const std::u32string str = u16string_to_u32string(u"e");
+    assert_equal(str.length(), 1);
+    assert_equal(str.at(0), 0x0065);
+  }
+  {
+    const std::u32string str = u16string_to_u32string(u"🙀");
+    assert_equal(str.length(), 1);
+    assert_equal(str.at(0), 0x0001F640);
+  }
+  {
+    const std::u32string str = u16string_to_u32string(u"Ω🙀");
+    assert_equal(str.length(), 2);
+    assert_equal(str.at(0), u'Ω');
+    assert_equal(str.at(1), 0x0001F640);
+  }
+
+  // error cases
+  {
+    std::u16string half_cat;
+    half_cat.push_back(0xD83D);  // mismatched lead surrogate
+    const std::u32string str = u16string_to_u32string(half_cat);
+    assert_equal(str.length(), 1);
+    assert_equal(str.at(0), 0xFFFD);
+  }
+  {
+    std::u16string half_cat;
+    half_cat.push_back(0xD83D);  // mismatched lead surrogate
+    half_cat.push_back(u'Ω');    // with following text
+    const std::u32string str = u16string_to_u32string(half_cat);
+    assert_equal(str.length(), 2);
+    assert_equal(str.at(0), 0xFFFD);
+    assert_equal(str.at(1), u'Ω');
+  }
+  {
+    std::u16string half_cat;
+    half_cat.push_back(0xDE40);  // mismatched trail surrogate
+    const std::u32string str = u16string_to_u32string(half_cat);
+    assert_equal(str.length(), 1);
+    assert_equal(str.at(0), 0xFFFD);
+  }
+}
```

Comment:
You could add another test:

```suggestion
    {
        std::u16string no_cat;
        no_cat.push_back(0xDE40);
        no_cat.push_back(0xD83D);
        const std::u32string str = u16string_to_u32string(no_cat);
        assert_equal(str.length(), 2);
        assert_equal(str.at(0), 0xFFFD);
        assert_equal(str.at(1), 0xFFFD);
    }
}
```

Reply:
it's a great suggestion, but this PR was incorrectly based, that code is in #7442 — I will add it there.

Reply:
split to PR https://github.com/keymanapp/keyman/pull/7454